client/asset/btc/electrum/network.go

```diff
@@ -0,0 +1,646 @@
+// This code is available on the terms of the project LICENSE.md file,
+// also available online at https://blueoakcouncil.org/license/1.0.0.
+
+package electrum
+
+import (
+	"bufio"
+	"context"
+	"crypto/tls"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"net"
+	"strconv"
+	"strings"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/decred/go-socks/socks"
+)
+
+const pingInterval = 10 * time.Second
+
+// ServerConn represents a connection to an Electrum server e.g. ElectrumX. It
+// is a single use type that must be replaced if the connection is lost. Use
+// ConnectServer to construct a ServerConn and connect to the server.
+type ServerConn struct {
+	conn   net.Conn
+	cancel context.CancelFunc
+	done   chan struct{}
+	proto  string
+
+	reqID uint64
+
+	respHandlersMtx sync.Mutex
+	respHandlers    map[uint64]chan *response // reqID => requestor
+
+	ntfnHandlersMtx sync.RWMutex
+	ntfnHandlers    map[string][]chan []byte // method => subscribers
+}
+
+func (sc *ServerConn) nextID() uint64 {
+	return atomic.AddUint64(&sc.reqID, 1)
+}
+
+const newline = byte('\n')
+
+func (sc *ServerConn) listen(ctx context.Context) {
+	defer sc.cancelRequests() // close the response chans
+	defer sc.deleteSubscriptions()
+
+	reader := bufio.NewReader(io.LimitReader(sc.conn, 1<<18))
+
+	for {
+		if ctx.Err() != nil {
+			return
+		}
+		msg, err := reader.ReadBytes(newline)
+		if err != nil {
+			if ctx.Err() == nil { // unexpected
+				fmt.Println(err)
+			}
+			sc.cancel()
+			return
+		}
+
+		var jsonResp response
+		err = json.Unmarshal(msg, &jsonResp)
+		if err != nil {
+			fmt.Println("Unmarshal error:", err)
+			continue
+		}
+
+		if jsonResp.Method != "" { // notification
+			var ntfnParams ntfnData // the ntfn payload is in the params field of a request object (!)
+			err = json.Unmarshal(msg, &ntfnParams)
+			if err != nil {
+				fmt.Println("Unmarshal error:", err)
+				continue
+			}
+			for _, c := range sc.subChans(jsonResp.Method) {
+				select {
+				case c <- ntfnParams.Params:
+				default:
+				}
+			}
+			continue
+		}
+		// fmt.Println(string(msg))
+
+		c := sc.responseChan(jsonResp.ID)
+		if c == nil {
+			fmt.Printf("Received response for unknown request ID %d", jsonResp.ID)
+			continue
+		}
+		c <- &jsonResp
+	}
+}
+
+func (sc *ServerConn) pinger(ctx context.Context) {
+	t := time.NewTicker(pingInterval)
+	defer t.Stop()
+
+	for {
+		// listen => ReadBytes cannot wait forever. Reset the read deadline for
+		// the next ping's response, as the ping loop is running.
+		err := sc.conn.SetReadDeadline(time.Now().Add(pingInterval * 5 / 4))
```

Comment:
It's global, just like with gorilla.  We design around that because when the last-set deadline is exceeded, all read operations fail instead of blocking, allowing the read loop to break as well.

If the ping takes over 5 seconds to get a response, we should cut off everything and fail the connection.

Reply:
So, setting the deadline also sets if for requests we are already waiting for?